src/metrics/rules/active_ruleset.go

```diff
@@ -144,46 +135,72 @@ func (as *activeRuleSet) ForwardMatch(
-	if err != nil {
-		return forwardMatchResult{}, err
+	matchOpts *MatchOptions,
+) error {
+	if err := as.mappingsForNonRollupID(id, timeNanos, matchOpts); err != nil {
+		return err
 	}
-	rollupResults, err := as.rollupResultsFor(id, timeNanos, matchOpts)
-	if err != nil {
-		return forwardMatchResult{}, err
+
+	if err := as.rollupResultsFor(id, timeNanos, matchOpts); err != nil {
+		return err
+	}
+
+	ensureNonDecreasingCutoverForExistingID(matchOpts.MatchResult, timeNanos)
+	dedupeForExistingIDStagedMetadatas(matchOpts.MatchResult)
+
+	return nil
+}
+
+func dedupeForExistingIDStagedMetadatas(matchResult *MatchResult) {
+	if len(matchResult.forExistingID) == 0 {
+		return
+	}
+
+	idx := len(matchResult.forExistingID) - 1
+	if len(matchResult.forExistingID[idx].Pipelines) == 0 {
+		return
+	}
+
+	// Otherwise merge as per usual
+	curr := 0
+	for i := 1; i < len(matchResult.forExistingID[idx].Pipelines); i++ {
+		foundDup := false
+		for j := 0; j <= curr; j++ {
+			if matchResult.forExistingID[idx].Pipelines[j].Equal(matchResult.forExistingID[idx].Pipelines[i]) {
+				foundDup = true
+				break
+			}
+		}
+		if foundDup {
+			continue
+		}
+		curr++
+		matchResult.forExistingID[idx].Pipelines[curr] = matchResult.forExistingID[idx].Pipelines[i]
 	}
-	forExistingID := mappingResults.forExistingID.
-		merge(rollupResults.forExistingID).
-		unique().
-		toStagedMetadata()
-	forNewRollupIDs := make([]IDWithMetadatas, 0, len(rollupResults.forNewRollupIDs))
-	for _, idWithMatchResult := range rollupResults.forNewRollupIDs {
-		stagedMetadata := idWithMatchResult.matchResults.unique().toStagedMetadata()
-		newIDWithMetadatas := IDWithMetadatas{
-			ID:        idWithMatchResult.id,
-			Metadatas: metadata.StagedMetadatas{stagedMetadata},
-		}
-		forNewRollupIDs = append(forNewRollupIDs, newIDWithMetadatas)
+	for i := curr + 1; i < len(matchResult.forExistingID[idx].Pipelines); i++ {
+		matchResult.forExistingID[idx].Pipelines[i] = metadata.PipelineMetadata{}
 	}
-	sort.Sort(IDWithMetadatasByIDAsc(forNewRollupIDs))
-	return forwardMatchResult{
-		forExistingID:   forExistingID,
-		forNewRollupIDs: forNewRollupIDs,
-		keepOriginal:    rollupResults.keepOriginal,
-	}, nil
+	matchResult.forExistingID[idx].Pipelines = matchResult.forExistingID[idx].Pipelines[:curr+1]
 }
 
 func (as *activeRuleSet) mappingsForNonRollupID(
 	id []byte,
 	timeNanos int64,
-	matchOpts MatchOptions,
-) (mappingResults, error) {
+	matchOpts *MatchOptions,
+) error {
 	var (
 		cutoverNanos int64
-		pipelines    []metadata.PipelineMetadata
+		matchResult  = matchOpts.MatchResult
 	)
+
+	if cap(matchResult.forExistingID) > len(matchResult.forExistingID) {
+		matchResult.forExistingID = matchResult.forExistingID[0 : len(matchResult.forExistingID)+1]
+	} else {
+		matchResult.forExistingID = append(matchResult.forExistingID, metadata.StagedMetadata{})
+	}
+	numMetadatas := len(matchResult.forExistingID)
+	idx := numMetadatas - 1
```

Comment:
By increasing the `forExistingID` StagedMetadatas array by 1 and using the last index for updates, we get rid of the need for https://github.com/m3db/m3/pull/4019/files#diff-2d6d0ddef3f98e8ba7f8d8a15c0367b893bd32dec1fcddf062c3c9034eca966bL121 as metadata does not need to be merged into the result later